Added error handling in regionprops_table with fill_value

ElpadoCan opened the pull request:
The user can now set `errors='ignore'` to `skimage.measure.regionprops_table` and, optionally, a `fill_value` to replace region properties that raised an error with `fill_value` (default `fill_value` is `None`). This allows getting a table with all of the properties that did not raise any exception, instead of raising an exception for the entire table.

<!--
Please use `pre-commit` to format code.

```
pip install pre-commit  # install the package
pre-commit install  # install git commit hook
```

Now, formatting checks will be run on each commit.
You can also run `pre-commit` manually:

```
pre-commit run -a
```
-->

## Description

Implementation of the feature that I requested in #6603. Note that if an error is raised on a specific property of all the regions and this property was a tuple or array, then the property will appear in the table with its non-modified name. This is because it was never computed, so it was never possible to determine if it was a scalar, tuple or array.

## Checklist

<!-- It's fine to submit PRs which are a work in progress! -->
<!-- But before they are merged, all PRs should provide: -->

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- Descriptive commit messages (see below)

<!-- For detailed information on these and other aspects see -->
<!-- the scikit-image contribution guidelines. -->
<!-- https://scikit-image.org/docs/dev/contribute.html -->

## For reviewers

<!-- Don't remove the checklist below. -->

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
  `doc/release/release_dev.rst`.
- There is a bot to help automate backporting a PR to an older branch. For
  example, to backport to v0.19.x after merging, add the following in a PR
  comment: `@meeseeksdev backport to v0.19.x`
- To run benchmarks on a PR, add the `run-benchmark` label. To rerun, the label
  can be removed and then added again. The benchmark output can be checked in
  the "Actions" tab.

ElpadoCan committed:
Added error handling in regionprops_table with fill_value

The user can now set `errors='ignore'` and a optionally a `fill_value` to
replace region properties that raised an error with `fill_value`.
This allows getting a table with all of the properties that did not raise
any exception, instead of raising an exception for the entire table

ElpadoCan committed:
removed line of code used for testing

ElpadoCan commented:
Tests failed because I inadvertently pushed with a line of code that I used to test. I removed it with a second commit, but I don't think I can run the tests again myself. Sorry about that

stefanv commented:
Thanks @ElpadoCan. I've had only a very quick look, but my sense is that this is overly complicated right now. Can we think of ways to simplify the implementation?

ElpadoCan commented:
Hi @stefanv, I agree the whole thing is a bit complicated, but there is a reason and I have some ideas on how to simplify it. 

The way `regionprops_table` currently works is by looping the properties and computing the property of the first region in the list of `regionprops`. This is done to determine whether the property is a scalar, tuple or array. However, if we want to ignore errors, when we get an error on the first region then we would not know the property type. This is why I added a small for loop (only for the case of `errors='ignore'`) that breaks as soon as the property doesn't raise errors. This is the for loop I added:

```python
# Try to compute `rp` for all objects to get one that doesn't raise
# exceptions
for _region in regions:
    try:
        rp = getattr(_region, prop)
        break
    except Exception as err:
        continue
else:
    # break was never called --> all objects raise exceptions
    rp = None
```

Now comes the second edge case: what if the property raises an exception on ALL regions (for loop never breaks, hence the `else` clause after the for loop)? We would not know the property type so we cannot add it to the dictionary yet. This is why I am saving this property name in a dictionary called `missing_props_info` and I add the missing properties at the end.

One way to simplify this would be to not cover the edge case of a property raising exceptions on ALL regions. This would simplify everything and for this rare case, we add a note in the documentation that, when this happens, the property will not appear as a column.

If we do this we only need the for loop I explained above and the `try/except` when we compute the property.

The other way would be to return a pandas DataFrame instead of a dictionary because pandas can automatically fill missing values with `nan` which we can replace with `fill_value` after all properties have been added to the table.

Let me know what you think and I can implement it, thanks!

ElpadoCan committed:
simplified original implementation

ElpadoCan committed:
removed trailing whitespace

ElpadoCan commented:
Ok, I just realised that we didn't need the two new functions I made so now the implementation is a lot simpler. We still cover the edge case of all regions raising an exception with a specific property by adding this property to the dictionary with its non-modified name (for non-scalar properties). 

Let me know what you think now, thanks!

ElpadoCan committed:
removed import re since it is not needed anymore

ElpadoCan committed:
removed missing_props_info since it is not needed anymore

ElpadoCan commented:
I still believe that this can be very useful 😁

lagru commented:
Agreed! And sorry for the hiatus. Reviewer time can be a bottle neck. :/  Do you mind looking into the merge conflict? I can review this afterwards. :)

lagru commented:
Oh and let us know if you have any questions.

ElpadoCan commented:
> Agreed! And sorry for the hiatus. Reviewer time can be a bottle neck. :/ Do you mind looking into the merge conflict? I can review this afterwards. :)

Absolutely no problem I fully understand. I will check the merge conflicts now, thanks!

ElpadoCan commented:
So the merge conflict is pretty easy, it's basically the addition of the code I added for this PR. Technically it should not be a conflict, but I think git complains because in the meantime two versions have been released. I would simply accept what is coming from my branch, but I don't think I have the power to do that, right?

ElpadoCan commented:
Hi everyone, any chance this can be merged? I talked to other people who also told me they would find this useful. Thanks!

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
